// @create-figma-plugin/ui/lib/components/layer/stories/layer-frame-selected.stories.js
import { h } from 'preact';
import { useState } from 'preact/hooks';
import { useInitialFocus } from '../../../hooks/use-initial-focus/use-initial-focus.js';
import { IconFrame16 } from '../../../icons/icon-16/icon-frame-16.js';
import { Layer } from '../layer.js';
export default {
    parameters: {
        fixedWidth: true
    },
    tags: ['2'],
    title: 'Components/Layer/Frame Selected'
};
export const Passive = function () {
    const [value, setValue] = useState(true);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { icon: h(IconFrame16, null), onChange: handleChange, value: value }, "Text"));
};
export const Focused = function () {
    const [value, setValue] = useState(true);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { ...useInitialFocus(), icon: h(IconFrame16, null), onChange: handleChange, value: value }, "Text"));
};
export const Bold = function () {
    const [value, setValue] = useState(true);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { bold: true, icon: h(IconFrame16, null), onChange: handleChange, value: value }, "Text"));
};
export const Description = function () {
    const [value, setValue] = useState(true);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { description: "Description", icon: h(IconFrame16, null), onChange: handleChange, value: value }, "Text"));
};
export const OnValueChange = function () {
    const [value, setValue] = useState(true);
    function handleValueChange(newValue) {
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { icon: h(IconFrame16, null), onValueChange: handleValueChange, value: value }, "Text"));
};
